[ISSUE] Issue with `databricks_sql_visualization` resource

Hi guys,

Like 6 months ago we had a requirement, deploy a dashboard in multiple workspaces. In that moment what I did was make the dashboard in one workspace and using this api: https://databricks-instance/api/2.0/preview/sql/dashboards/{dashboardId} I "exported" it as a JSON, and using this JSON I mapped my Terraform script to deploy on the other workspaces.

That worked fine except for the visualizations because when the visualization wasn't a table, it looked like this:
<img width="697" alt="Screen Shot 2022-11-10 at 6 55 23 PM" src="https://user-images.githubusercontent.com/16740850/201230500-92d31f32-245f-4afa-b1f5-4e1a62cd0f81.png">
This issue only affects the chart in the dashboard because in the query editor looks perfect.

In that moment we contacted Databricks support and after a while and testing also with this Python tool: https://github.com/QuentinAmbard/databricks-sql-clone we realized that the problem was the query_plan field (it's required for this purpose), so in conclusion, we need to pass the query_plan field to the API.

In this commit https://github.com/QuentinAmbard/databricks-sql-clone/commit/e739329a2c21604183741006a677522d56dcc8c6 you can see that this Python tool added the support to the query_plan field.

In that moment, the temporal solution was edit (only open and save) each visualization in each workspace manually! (I dont understant how that works but I suppose is like a refresh) and it was a lot of work because in that moment we had 20+ workspaces.

Now, I have to add new charts to this dashboard and for my surprise, this feature is not supported :(

### Configuration
```hcl
# This is a test Tf script

locals {
    dashboard = jsondecode(file("dashboard.json"))
    widget_data = local.dashboard.widgets[0]
}

provider "databricks" {
  host = "<workspace-url>"
  token = "<token>"
}

# Endpoint
resource "databricks_sql_endpoint" "test_endpoint" {
  name             = "test"
  cluster_size     = "2X-Small"
  auto_stop_mins   = 10
  min_num_clusters = 1
  max_num_clusters = 1
  num_clusters     = 0
  warehouse_type   = "PRO"
}

# Dashboard
resource "databricks_sql_dashboard" "test_dashboard" {
  name = local.dashboard.name
  tags = local.dashboard.tags
}

# Dashboard items
resource "databricks_sql_query" "test_query" {
  name           = local.widget_data.visualization.query.name
  query          = local.widget_data.visualization.query.query
  data_source_id = databricks_sql_endpoint.test_endpoint.data_source_id
  run_as_role    = "owner"
}

resource "databricks_sql_visualization" "test_visu" {
  query_id    = databricks_sql_query.test_query.id
  type        = lower(local.widget_data.visualization.type)
  name        = local.widget_data.visualization.name
  description = local.widget_data.visualization.description

  options = jsonencode(local.widget_data.visualization.options)
  #query_plan = jsonencode(local.widget_data.visualization.query_plan)
}

resource "databricks_sql_widget" "test_widget" {
  dashboard_id      = databricks_sql_dashboard.test_dashboard.id
  visualization_id  = databricks_sql_visualization.test_visu.id
  title             = local.widget_data.options.title
  description       = local.widget_data.options.description

  position {
    size_x = local.widget_data.options.position.sizeX
    size_y = local.widget_data.options.position.sizeY
    pos_x  = local.widget_data.options.position.col
    pos_y  = local.widget_data.options.position.row
  }
}
```
this is my `dashboard.json` file, I just left the most important fields.
``` hcl
{    
    "slug": "dashboard_test",
    "name": "dashboard_test",
    "layout": [],
    "widgets": [
        {
            "width": 1,
            "options": {
                "parameterMappings": {},
                "title": "test widgets",
                "description": "",
                "isHidden": false,
                "position": {
                    "autoHeight": false,
                    "sizeX": 3,
                    "sizeY": 9,
                    "minSizeX": 1,
                    "maxSizeX": 6,
                    "minSizeY": 5,
                    "maxSizeY": 1000,
                    "col": 0,
                    "row": 30
                },
                "overrideColors": true
            },
            "text": "",
            "updated_at": "2022-05-10T16:03:59Z",
            "created_at": "2022-04-29T18:59:35Z",
            "visualization": {                
                "type": "CHART",
                "name": "Bar 2",
                "description": "",
                "options": {
                    "version": 2,
                    "globalSeriesType": "column",
                    "sortX": true,
                    "sortY": true,
                    "legend": {
                        "traceorder": "normal"
                    },
                    "xAxis": {
                        "type": "-",
                        "labels": {
                            "enabled": true
                        },
                        "title": {
                            "text": "Date"
                        }
                    },
                    "yAxis": [
                        {
                            "type": "-",
                            "title": {
                                "text": "Total Jobs"
                            }
                        },
                        {
                            "type": "-",
                            "opposite": true,
                            "title": {
                                "text": null
                            }
                        }
                    ],
                    "alignYAxesAtZero": true,
                    "error_y": {
                        "type": "data",
                        "visible": true
                    },
                    "series": {
                        "stacking": null,
                        "error_y": {
                            "type": "data",
                            "visible": true
                        }
                    },
                    "seriesOptions": {
                        "column_9bd3d92f388": {
                            "yAxis": 0,
                            "type": "column"
                        },
                        "Failed": {
                            "color": "#E92828"
                        },
                        "Succeeded": {
                            "color": "#3BD973"
                        }
                    },
                    "valuesOptions": {},
                    "direction": {
                        "type": "counterclockwise"
                    },
                    "sizemode": "diameter",
                    "coefficient": 1,
                    "numberFormat": "0,0[.]00000",
                    "percentFormat": "0[.]00%",
                    "textFormat": "",
                    "missingValuesAsZero": true,
                    "useAggregationsUi": true,
                    "swappedAxes": false,
                    "dateTimeFormat": "YYYY-MM-DD HH:mm",
                    "showDataLabels": false,
                    "columnConfigurationMap": {
                        "x": {
                            "column": "day",
                            "id": "column_9bd3d92f387"
                        },
                        "y": [
                            {
                                "column": "Count",
                                "transform": "SUM",
                                "id": "column_9bd3d92f388"
                            }
                        ],
                        "series": {
                            "column": "State",
                            "id": "column_9bd3d92f389"
                        }
                    }
                },
                "updated_at": "2022-05-11T01:21:04Z",
                "created_at": "2022-04-28T14:53:42Z",
                "query_plan": {
                    "selects": [
                        {
                            "column": "day"
                        },
                        {
                            "function": "SUM",
                            "args": [
                                {
                                    "column": "Count"
                                }
                            ],
                            "alias": "column_9bd3d92f388"
                        },
                        {
                            "column": "State"
                        }
                    ],
                    "groups": [
                        {
                            "column": "day"
                        },
                        {
                            "column": "State"
                        }
                    ]
                },
                "query": {                    
                    "name": "Super awesome query",
                    "description": "",
                    "query": "select\n  to_date(job_runtime.startTS) as day,\n  job_terminal_state as State,\n  count(1) as Count\nfrom\n  overwatch.jobrun\ngroup by\n  to_date(job_runtime.startTS),\n  job_terminal_state\nhaving\n  day > date_sub(current_date(), 30)\norder by\n  day desc",
                    "schedule": null,
                    "is_archived": false,
                    "is_draft": false,
                    "updated_at": "2022-05-09T23:38:10Z",
                    "created_at": "2022-04-28T02:15:36Z",                    
                    "options": {
                        "run_as_role": "owner",
                        "apply_auto_limit": false,
                        "parameters": []
                    },
                    "version": 1,
                    "tags": [],
                    "is_safe": true                    
                }
            }
        }        
    ],
    "is_archived": false,
    "is_draft": false,
    "tags": ["test"]
}
```


### Steps to Reproduce
This issue only happens when the dashboard/visualization is deployed in ANOTHER workspace, if the same workspace is used to "export" and "import" nothing happens. So I'm sure if you use the values from my `dashboard.json` file, you'll face the issue.


1. create a `main.tf`file with the code above and set the `host`and `token`values.
2. create the `dashboard.json` with the JSON above at the same location that `main.tf`
3. just execute terraform as usually -> `terraform plan` and `terraform apply -auto-approve`

### Terraform and provider versions
query_plan field is not supported by any provider version, so you can use whatever you want. The configuration that we have is:
- Terraform:0.13
- terraform-provider-databricks:latest


### Important Factoids
I also tried to create the visualization through the Databricks API "https://databricks-instance/api/2.0/preview/sql/visualizations" adding the query_plan field and it worked (you can use the the JSON above for the body part) so the conclusion of all of this is the Databricks Visualization API supports this field, the problem is just here in the provider.

With this in mind, I tried to fix this by myself but I dont know why the `make coverage` command fails so I cant run the tests. I got this error but it's not important or related to the issue:
``` hcl
make: staticcheck: No such file or directory
make: *** [lint] Error 1
```

Anyway, this is the fork I made with the changes https://github.com/fava9410/terraform-provider-databricks I ran it and it worked but for some reason when I ran again `terraform plan` it detects new changes in the query_plan field (I think the changes are some whitespaces but I dont know how to handle this)

--- a/sql/acceptance/sql_dashboard_test.go
+++ b/sql/acceptance/sql_dashboard_test.go
@@ -69,7 +69,12 @@ func TestAccDashboard(t *testing.T) {
 					# Note: this resource differs from the one above in that
 					# the query plan is set. This tests that it can either
 					# be unset or set and in both cases yield a consistent result.
-					query_plan = jsonencode({})
+					query_plan = jsonencode({
+						# The value should be non-empty to check we don't have
+						# plan changes due to whitespace changes in JSON serialization.
+						groups = [
+						]
+					})
 				}
 			`,
 		},

--- a/sql/acceptance/sql_dashboard_test.go
+++ b/sql/acceptance/sql_dashboard_test.go
@@ -57,6 +57,7 @@ func TestAccDashboard(t *testing.T) {
 					name = "My Table"
 
 					options = jsonencode({})
+					query_plan = jsonencode({})
 				}
 			`,
 		},

--- a/sql/acceptance/sql_dashboard_test.go
+++ b/sql/acceptance/sql_dashboard_test.go
@@ -57,6 +57,18 @@ func TestAccDashboard(t *testing.T) {
 					name = "My Table"
 
 					options = jsonencode({})
+				}
+
+				resource "databricks_sql_visualization" "q1v2" {
+					query_id = databricks_sql_query.q1.id
+					type = "table"
+					name = "My Table (1)"
+
+					options = jsonencode({})
+
+					# Note: this resource differs from the one above in that
+					# the query plan is set. This tests that it can either
+					# be unset or set and in both cases yield a consistent result.
 					query_plan = jsonencode({})
 				}
 			`,

--- a/sql/api/visualization.go
+++ b/sql/api/visualization.go
@@ -13,4 +13,5 @@ type Visualization struct {
 	Name        string          `json:"name"`
 	Description string          `json:"description,omitempty"`
 	Options     json.RawMessage `json:"options,omitempty"`
+	QueryPlan   json.RawMessage `json:"query_plan,omitempty"`
 }

--- a/sql/api/visualization_test.go
+++ b/sql/api/visualization_test.go
@@ -15,6 +15,7 @@ func TestVisualizationMarshalUnmarshal(t *testing.T) {
 		Name:        "name",
 		Description: "description",
 		Options:     json.RawMessage("{}"),
+		QueryPlan:   json.RawMessage(`{"foo":"bar"}`),
 	}
 
 	out, err := json.Marshal(v)

--- a/sql/resource_sql_visualization.go
+++ b/sql/resource_sql_visualization.go
@@ -134,26 +134,28 @@ func jsonRemarshal(in []byte) ([]byte, error) {
 	return out, nil
 }
 
+func suppressWhitespaceChangesInJSON(_, old, new string, d *schema.ResourceData) bool {
+	oldp, err := jsonRemarshal([]byte(old))
+	if err != nil {
+		log.Printf("[WARN] Unable to remarshal value %#v", old)
+		return false
+	}
+	newp, err := jsonRemarshal([]byte(new))
+	if err != nil {
+		log.Printf("[WARN] Unable to remarshal value %#v", new)
+		return false
+	}
+	return bytes.Equal(oldp, newp)
+}
+
 func ResourceSqlVisualization() *schema.Resource {
 	p := common.NewPairSeparatedID("query_id", "visualization_id", "/")
 	s := common.StructToSchema(
 		VisualizationEntity{},
 		func(m map[string]*schema.Schema) map[string]*schema.Schema {
-			// We care only about logical changes to the JSON payload in `options`.
-			m["options"].DiffSuppressFunc = func(_, old, new string, d *schema.ResourceData) bool {
-				oldp, err := jsonRemarshal([]byte(old))
-				if err != nil {
-					log.Printf("[WARN] Unable to remarshal value %#v", old)
-					return false
-				}
-				newp, err := jsonRemarshal([]byte(new))
-				if err != nil {
-					log.Printf("[WARN] Unable to remarshal value %#v", new)
-					return false
-				}
-				return bytes.Equal(oldp, newp)
-			}
-
+			// We care only about logical changes to the JSON payload in `options` and `query_plan`.
+			m["options"].DiffSuppressFunc = suppressWhitespaceChangesInJSON
+			m["query_plan"].DiffSuppressFunc = suppressWhitespaceChangesInJSON
 			return m
 		})
 

--- a/sql/resource_sql_visualization.go
+++ b/sql/resource_sql_visualization.go
@@ -23,6 +23,7 @@ type VisualizationEntity struct {
 	Name        string `json:"name"`
 	Description string `json:"description,omitempty"`
 	Options     string `json:"options"`
+	QueryPlan   string `json:"query_plan,omitempty"`
 }
 
 func (v *VisualizationEntity) toAPIObject(schema map[string]*schema.Schema, data *schema.ResourceData) (*api.Visualization, error) {
@@ -37,6 +38,9 @@ func (v *VisualizationEntity) toAPIObject(schema map[string]*schema.Schema, data
 	av.Name = v.Name
 	av.Description = v.Description
 	av.Options = json.RawMessage(v.Options)
+	if v.QueryPlan != "" {
+		av.QueryPlan = json.RawMessage(v.QueryPlan)
+	}
 	return &av, nil
 }
 
@@ -48,6 +52,9 @@ func (v *VisualizationEntity) fromAPIObject(av *api.Visualization, schema map[st
 	v.Name = av.Name
 	v.Description = av.Description
 	v.Options = string(av.Options)
+	if av.QueryPlan != nil {
+		v.QueryPlan = string(av.QueryPlan)
+	}
 
 	// Transform to ResourceData.
 	return common.StructToData(*v, schema, data)

--- a/sql/resource_sql_visualization.go
+++ b/sql/resource_sql_visualization.go
@@ -52,7 +52,9 @@ func (v *VisualizationEntity) fromAPIObject(av *api.Visualization, schema map[st
 	v.Name = av.Name
 	v.Description = av.Description
 	v.Options = string(av.Options)
-	if av.QueryPlan != nil {
+
+	// If the query plan attribute is not set; it may come back as `null` from the API.
+	if av.QueryPlan != nil && !bytes.Equal(av.QueryPlan, []byte("null")) {
 		v.QueryPlan = string(av.QueryPlan)
 	}
 

--- a/sql/resource_sql_visualization_test.go
+++ b/sql/resource_sql_visualization_test.go
@@ -183,16 +183,58 @@ func TestVisualizationReadWithQueryPlan(t *testing.T) {
 		},
 	}.Apply(t)
 
-	assert.NoError(t, err, err)
-
-	assert.Equal(t, "foo/12345", d.Id())
-	assert.Equal(t, "foo", d.Get("query_id"))
-	assert.Equal(t, "12345", d.Get("visualization_id"))
-	assert.Equal(t, "chart", d.Get("type"))
-	assert.Equal(t, "My Chart", d.Get("name"))
-	assert.Equal(t, "Some Description", d.Get("description"))
-	assert.Less(t, 0, len(d.Get("options").(string)))
+	assert.NoError(t, err)
 	assert.Equal(t, `{"foo":"qux"}`, d.Get("query_plan").(string))
+}
+
+func TestVisualizationReadWithNullQueryPlan(t *testing.T) {
+	d, err := qa.ResourceFixture{
+		Fixtures: []qa.HTTPFixture{
+			{
+				Method:   "GET",
+				Resource: "/api/2.0/preview/sql/queries/foo",
+				Response: api.Query{
+					ID: "foo",
+					Visualizations: []json.RawMessage{
+						json.RawMessage(`
+							{
+								"id": 12344
+							}
+						`),
+						json.RawMessage(`
+							{
+								"id": 12345,
+								"type": "CHART",
+								"name": "My Chart",
+								"description": "Some Description",
+								"options": {},
+								"query_plan": null
+							}
+						`),
+						json.RawMessage(`
+							{
+								"id": 12345
+							}
+						`),
+					},
+				},
+			},
+		},
+		Resource: ResourceSqlVisualization(),
+		Read:     true,
+		New:      true,
+		ID:       "foo/12345",
+		State: map[string]any{
+			"query_id":    "foo",
+			"type":        "chart",
+			"name":        "My Chart",
+			"description": "Some Description",
+			"options":     "{}",
+		},
+	}.Apply(t)
+
+	assert.NoError(t, err)
+	assert.Equal(t, "", d.Get("query_plan").(string))
 }
 
 func TestVisualizationReadNotFound(t *testing.T) {

--- a/sql/resource_sql_visualization_test.go
+++ b/sql/resource_sql_visualization_test.go
@@ -137,6 +137,64 @@ func TestVisualizationRead(t *testing.T) {
 	assert.Less(t, 0, len(d.Get("options").(string)))
 }
 
+func TestVisualizationReadWithQueryPlan(t *testing.T) {
+	d, err := qa.ResourceFixture{
+		Fixtures: []qa.HTTPFixture{
+			{
+				Method:   "GET",
+				Resource: "/api/2.0/preview/sql/queries/foo",
+				Response: api.Query{
+					ID: "foo",
+					Visualizations: []json.RawMessage{
+						json.RawMessage(`
+							{
+								"id": 12344
+							}
+						`),
+						json.RawMessage(`
+							{
+								"id": 12345,
+								"type": "CHART",
+								"name": "My Chart",
+								"description": "Some Description",
+								"options": {},
+								"query_plan": {"foo":"qux"}
+							}
+						`),
+						json.RawMessage(`
+							{
+								"id": 12345
+							}
+						`),
+					},
+				},
+			},
+		},
+		Resource: ResourceSqlVisualization(),
+		Read:     true,
+		ID:       "foo/12345",
+		State: map[string]any{
+			"query_id":    "foo",
+			"type":        "chart",
+			"name":        "My Chart",
+			"description": "Some Description",
+			"options":     "{}",
+			"query_plan":  `{"foo":"bar"}`,
+		},
+	}.Apply(t)
+
+	assert.NoError(t, err, err)
+
+	assert.Equal(t, "foo/12345", d.Id())
+	assert.Equal(t, "foo", d.Get("query_id"))
+	assert.Equal(t, "12345", d.Get("visualization_id"))
+	assert.Equal(t, "chart", d.Get("type"))
+	assert.Equal(t, "My Chart", d.Get("name"))
+	assert.Equal(t, "Some Description", d.Get("description"))
+	assert.Less(t, 0, len(d.Get("options").(string)))
+	assert.Equal(t, `{"foo":"qux"}`, d.Get("query_plan").(string))
+}
+
 func TestVisualizationReadNotFound(t *testing.T) {
 	d, err := qa.ResourceFixture{
 		Fixtures: []qa.HTTPFixture{

